rust/src/utils/kmeans.rs

```diff
@@ -159,33 +167,51 @@ struct KMeanMembership {
     /// Distance between each vector, to its corresponding centroids.
     distances: Vec<f32>,
 
+    /// Number of centroids.
     k: u32,
 }
 
 impl KMeanMembership {
+    /// Reconstruct a KMeans model from the membership.
     async fn to_kmean(&self) -> Result<KMeans> {
         let dimension = self.data.value_length();
         let cluster_ids = Arc::new(self.cluster_ids.clone());
         let data = self.data.clone();
+        let previous_centroids = self.centroids.clone();
         // New centroids for each cluster
         let means = stream::iter(0..self.k)
-            .zip(repeat_with(|| (data.clone(), cluster_ids.clone())))
-            .map(|(cluster, (data, cluster_ids))| async move {
-                tokio::task::spawn_blocking(move || {
-                    let mut sum = Float32Array::from_iter_values(
-                        (0..dimension).map(|_| 0.0).collect::<Vec<_>>(),
-                    );
-                    let mut total = 0.0;
-                    for i in 0..cluster_ids.len() {
-                        if cluster_ids[i] == cluster {
-                            sum = add(&sum, as_primitive_array(data.value(i).as_ref())).unwrap();
-                            total += 1.0;
-                        };
-                    }
-                    divide_scalar(&sum, total).unwrap()
-                })
-                .await
-            })
+            .zip(repeat_with(|| {
+                (
+                    data.clone(),
+                    cluster_ids.clone(),
+                    previous_centroids.clone(),
+                )
+            }))
+            .map(
+                |(cluster, (data, cluster_ids, previous_centroids))| async move {
+                    tokio::task::spawn_blocking(move || {
+                        let mut sum = Float32Array::from_iter_values(
+                            (0..dimension).map(|_| 0.0).collect::<Vec<_>>(),
+                        );
+                        let mut total = 0.0;
+                        for i in 0..cluster_ids.len() {
+                            if cluster_ids[i] == cluster {
```

Comment:
can you use the same binary_search speed up?

Reply:
This is `cluster_id.filter(where id = cluster)` which is `O(n)` regardless? Btw, the cluster_ids is ordered by the vector , not sorted by value